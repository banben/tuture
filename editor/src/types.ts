export type ExplainType = 'pre' | 'post';

export type EditMode = 'edit' | 'preview';
