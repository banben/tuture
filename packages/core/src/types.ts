import { Element } from 'tuture-slate';

import { DiffFile } from './diff';

export interface RawDiff {
  commit: string;
  diff: DiffFile[];
}

export interface Explain extends Element {
  type: 'explain';
  fixed: true;
}

export interface DiffBlock extends Element {
  type: 'diff-block';
  file: string;
  commit: string;
  hiddenLines?: [number, number][];
}

export interface File extends Element {
  type: 'file';
  file: string;
  display?: boolean;
  children: [Explain, DiffBlock, Explain];
}

export interface Meta {
  name: string;
  description?: string;
  id: string;
  cover?: string;
  created?: Date;
  topics?: string[];
  categories?: string[];
  github?: string;
}

export interface Article extends Meta {}

export interface StepTitle extends Element {
  type: 'heading-two';
  commit: string;
  id: string;
  fixed: true;
}

export type StepChild = StepTitle | Explain | File;

export interface Step extends Element {
  type: 'step';
  id: string;
  articleId?: string | null;
  outdated?: boolean;
  commit: string;
  children: StepChild[];
}

export interface Remote {
  name: string;
  refs: {
    fetch: string;
    push: string;
  };
}

export interface Collection extends Meta {
  version?: string;
  articles: Article[];
  remotes?: Remote[];
  steps: Step[];
}

export interface TutureConfig {
  ignoredFiles: string[];
  port: number;
}
