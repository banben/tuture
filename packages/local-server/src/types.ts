export interface TocArticleItem {
  type: 'article';
  id: string;
  name: string;
  level: number;
}

export interface TocStepItem {
  type: 'step';
  id: string;
  name: string;
  level: number;
  number: number;
  articleId?: string | null;
  outdated?: boolean;
}

export type TocItem = TocArticleItem | TocStepItem;

export interface CollectionStep {
  key: string;
  id: string;
  title: string;
  articleId?: string | null;
  articleIndex: number;
  articleName: string;
}
