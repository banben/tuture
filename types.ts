export interface Explain {
  pre?: string;
  post?: string;
}

export interface Diff {
  file: string;
  explain?: Explain;
  display?: boolean;
}

export interface TutureMeta {
  name: string;
  topics?: string[];
  categories?: string[];
  description?: string;
  created?: Date;
  updated?: Date;
  cover?: string;
  github?: string;
}

export interface Split extends TutureMeta {
  start: string;
  end: string;
}

export interface Commit {
  name: string;
  commit: string;
}

export interface Step extends Commit {
  explain?: Explain;
  diff: Diff[];
  outdated?: boolean;
}

export interface Tuture extends TutureMeta {
  id?: string;
  splits?: Split[];
  steps: Step[];
}

export interface TutureConfig {
  ignoredFiles: string[];
  port: number;
}
