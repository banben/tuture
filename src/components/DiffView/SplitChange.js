import React, { PureComponent } from 'react';
import classnames from 'classnames';

import {
  computeOldLineNumber,
  computeNewLineNumber,
} from './utils';

export default class SplitChange extends PureComponent {
  SIDE_OLD = 0;
  SIDE_NEW = 1;

  renderCells = ({
    side,
    change,
    gutterAnchor,
    hideGutter,
    gutterAnchorTarget
  }) => {
    if (!change) {
      const gutterClassName = classnames('diff-gutter', 'diff-gutter-omit');
      const codeClassName = classnames('diff-code', 'diff-code-omit');

      return [
        !hideGutter && <td key="gutter" className={gutterClassName} />,
        <td key="code" className={codeClassName} />,
      ];
    }

    const { type, content } = change;
    const line = side === this.SIDE_OLD
      ? computeOldLineNumber(change)
      : computeNewLineNumber(change)
    ;
    const gutterClassName = classnames(
      'diff-gutter',
      `diff-gutter-${type}`,
    );
    const gutterProps = {
      className: gutterClassName,
      'data-line-number': line,
      children: (
        gutterAnchor
        ? <a href={`#${gutterAnchorTarget}`} data-line-number={line} />
        : null
      ),
    };
    const codeClassName = classnames(
      'diff-code',
      `diff-code-${type}`,
    );
    const codeProps = {
      className: codeClassName,
    };

    return [
      !hideGutter && <td key="gutter" {...gutterProps} />,
      <td key="code" {...codeProps}>{content}</td>
    ];
  }

  render() {
    const {
      oldChange,
      newChange,
    } = this.props;

    const oldArgs = {
      change: oldChange,
      side: this.SIDE_OLD,
    };
    const newArgs = {
      change: newChange,
      side: this.SIDE_NEW,
    };

    const lineTypeClassName = ((oldChange, newChange) => {
      if (oldChange && !newChange) {
        return 'diff-line-old-only';
      }

      if (!oldChange && newChange) {
        return 'diff-line-new-only';
      }

      if (oldChange === newChange) {
        return 'diff-line-normal';
      }

      return 'diff-line-compare';
    })(oldChange, newChange);

    return (
      <tr className={classnames('diff-line')}>
        {this.renderCells(oldArgs)}
        {this.renderCells(newArgs)}
      </tr>
    );
  }
}
