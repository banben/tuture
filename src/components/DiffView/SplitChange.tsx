import React, { PureComponent } from 'react';
import classnames from 'classnames';

import {
  computeOldLineNumber,
  computeNewLineNumber,
} from './utils';

import { Change as ChangeType } from '../ContentItem';

export interface SplitChangeProps {
  oldChange: ChangeType;
  newChange: ChangeType;
}

export interface Cell {
  side?: number;
  change?: ChangeType;
  gutterAnchor?: string;
  hideGutter?: boolean;
  gutterAnchorTarget?: string;
  anchorID?: number;
  lineNumber?: number;
  hide?: boolean;
  className?: string;
}

export default class SplitChange extends PureComponent<SplitChangeProps> {
  constructor(props: SplitChangeProps) {
    super(props);
  }

  SIDE_OLD = 0;
  SIDE_NEW = 1;

  renderCells = ({
    side,
    change,
    gutterAnchor,
    hideGutter,
    gutterAnchorTarget
  }: Cell): React.ReactNode => {
    if (!change) {
      const gutterClassName = classnames('diff-gutter', 'diff-gutter-omit');
      const codeClassName = classnames('diff-code', 'diff-code-omit');

      return [
        !hideGutter && <td key="gutter" className={gutterClassName} />,
        <td key="code" className={codeClassName} />,
      ];
    }

    const { type, content } = change;
    const line = side === this.SIDE_OLD
      ? computeOldLineNumber(change)
      : computeNewLineNumber(change)
    ;
    const gutterClassName = classnames(
      'diff-gutter',
      `diff-gutter-${type}`,
    );
    const gutterProps = {
      className: gutterClassName,
      'data-line-number': line,
      children: (
        gutterAnchor
        ? <a href={`#${gutterAnchorTarget}`} data-line-number={line} />
        : null
      ),
    };
    const codeClassName = classnames(
      'diff-code',
      `diff-code-${type}`,
    );
    const codeProps = {
      className: codeClassName,
    };

    return [
      !hideGutter && <td key="gutter" {...gutterProps} />,
      <td key="code" {...codeProps}>{content}</td>
    ];
  }

  render() {
    const {
      oldChange,
      newChange,
    } = this.props;

    const oldArgs = {
      change: oldChange,
      side: this.SIDE_OLD,
    };
    const newArgs = {
      change: newChange,
      side: this.SIDE_NEW,
    };

    const lineTypeClassName = ((oldChange, newChange) => {
      if (oldChange && !newChange) {
        return 'diff-line-old-only';
      }

      if (!oldChange && newChange) {
        return 'diff-line-new-only';
      }

      if (oldChange === newChange) {
        return 'diff-line-normal';
      }

      return 'diff-line-compare';
    })(oldChange, newChange);

    return (
      <tr className={classnames('diff-line')}>
        {this.renderCells(oldArgs)}
        {this.renderCells(newArgs)}
      </tr>
    );
  }
}
