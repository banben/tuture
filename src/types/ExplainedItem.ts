type ExplainType = 'pre' | 'post';
type EditMode = 'edit' | 'preview';

type ToolType =
  | 'b'
  | 'i'
  | 'h'
  | 'list'
  | 'numbered list'
  | 'blockquotes'
  | 'link'
  | 'img'
  | 'code'
  | 'block code';

export { ExplainType, EditMode, ToolType };
