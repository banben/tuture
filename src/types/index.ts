interface Diff {
  file: string;
  explain: string;
  collapse?: boolean;
}

interface Step {
  name: string;
  commit: string;
  explain: string;
  diff: Diff[];
}

interface Tuture {
  name: string;
  language: string;
  version: string;
  topics: string[];
  description: string;
  maintainer: string;
  steps: Step[];
}

interface Change {
  content: string;
  type: string;
  isInsert?: boolean;
  isNormal?: boolean;
  isDelete?: boolean;
  lineNumber?: number;
  oldLineNumber?: number;
  newLineNumber?: number;
}

interface Hunk {
  content: string;
  newLines: number;
  oldLines: number;
  oldStart: number;
  newStart: number;
  changes: Change[];
  isPlain?: boolean;
}

export {
  Diff,
  Step,
  Tuture,
  Change,
  Hunk,
}
