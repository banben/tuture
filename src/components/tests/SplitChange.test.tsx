import React from 'react';
import renderer from 'react-test-renderer';

import SplitChange from '../DiffView/SplitChange';
import { Change } from '../../types/index';

test('SplitChange render successfully', () => {
  const oldChange: Change = {
    type: 'isdelete',
    content: 'I am the coolest one',
    isInsert: true,
    lineNumber: 0,
  };
  const newChange: Change = {
    type: 'isinsert',
    content: 'Come and play it',
    isInsert: true,
    lineNumber: 1,
  };
  const component = renderer.create(
    <SplitChange
      oldChange={oldChange}
      newChange={newChange}
      key={'test'}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
