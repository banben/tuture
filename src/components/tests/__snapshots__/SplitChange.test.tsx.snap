// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SplitChange render successfully 1`] = `
<tr
  className="diff-line"
>
  <td
    className="diff-gutter diff-gutter-isdelete"
    data-line-number={-1}
  />
  <td
    className="diff-code diff-code-isdelete"
  >
    I am the coolest one
  </td>
  <td
    className="diff-gutter diff-gutter-isinsert"
    data-line-number={1}
  />
  <td
    className="diff-code diff-code-isinsert"
  >
    Come and play it
  </td>
</tr>
`;
